docs/smwavset.1
.\" Automatically generated by Pandoc 3.1.3
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "smwavset" "1" "2026-01-06" "1.0.0-beta2" "SpectMorph Manual"
.hy
.SH NAME
.PP
smwavset - Program to manage spectmorph multiwave files
.SH SYNOPSIS
.PP
\f[B]smwavset\f[R] \f[I]\f[VI]<command>\f[I]\f[R] [\f[I]OPTIONS\f[R]] [
\f[I]\f[VI]<command specific args>\f[I]...\f[R] ]
.SH DESCRIPTION
.PP
\f[B]smwavset\f[R] is a command line tool to manage multi-wave files for
SpectMorph.
Ususally, a multi wave file is first initialized (using the
\f[B]init\f[R] command), then different wave files are added using the
\f[B]add\f[R] command.
After this has been done, encoding all waves can be done with one single
command, like decoding all waves.
.SH OPTIONS
.PP
\f[B]smwavset\f[R] follows the usual GNU command line syntax, with long
options starting with two dashes (\[aq]-\[aq]).
.TP
\f[B]-h\f[R], \f[B]--help\f[R]
Shows a brief help message.
.TP
\f[B]-v\f[R], \f[B]--version\f[R]
Prints out smwavset version.
.TP
\f[B]--args\f[R] \f[I]\f[VI]<arguments>\f[I]\f[R]
Specify additional arguments to be passed to smenc/smplay.
This could be --args \[dq]-O1\[dq] for encoding, or --args
\[dq]--no-sines\[dq] for decoding.
.TP
\f[B]-d\f[R], \f[B]--data-dir\f[R]
For encoding/decoding, this sets a data directory where the .wav files
or a list of .wav files (or .sm files), it is important that different
wavsets store their files in different directories (otherwise they will
overwrite wav files that belong to a different wavset).
So specifying a proper data directory which is different for each wavset
is usually necessary.
The default value is /tmp, which is only appropriate for testing; you
should ensure that scripts setup a different value.
.TP
\f[B]--format\f[R] \f[I]\f[VI]<field1>\f[I]\f[R],\f[I]\f[VI]<field2>\f[I]\f[R],\f[I]\f[VI]<field3>\f[I]\f[R],...,\f[I]\f[VI]<fieldN>\f[I]\f[R]
Set the output format for the list command.
.TP
\f[B]-j\f[R] \f[I]\f[VI]<jobs>\f[I]\f[R]
Use \f[I]\f[VI]<jobs>\f[I]\f[R] parallel jobs for encoding (for systems
with more than one processor).
.SH COMMANDS
.TP
\f[B]init\f[R] [ \f[I]\f[VI]<options>\f[I]\f[R] ] \f[I]\f[VI]<wavset>\f[I]\f[R]...
Initializes a new wavset; can also initialize more than one wavset
specified on the commandline.
.TP
\f[B]add\f[R] [ \f[I]\f[VI]<options>\f[I]\f[R] ] \f[I]\f[VI]<wavset>\f[I]\f[R] \f[I]\f[VI]<midi_note>\f[I]\f[R] \f[I]\f[VI]<path>\f[I]\f[R]
Adds a wave file to the wavset, where midi_note is the midi note the
wave file has been recorded for.
.TP
\f[B]list\f[R] [ \f[I]\f[VI]<options>\f[I]\f[R] ] \f[I]\f[VI]<wavset>\f[I]\f[R]
Lists the wave files that are contained within the wavset.
The output format for the list command can be specified using the format
option (comma seperated fields).
See FIELDS section for a list of valid fields
.TP
\f[B]encode\f[R] [ \f[I]\f[VI]<options>\f[I]\f[R] ] \f[I]\f[VI]<wset_filename>\f[I]\f[R] \f[I]\f[VI]<smset_filename>\f[I]\f[R]
Encodes a wavset using smenc.
.TP
\f[B]decode\f[R] [ \f[I]\f[VI]<options>\f[I]\f[R] ] \f[I]\f[VI]<smset_filename>\f[I]\f[R] \f[I]\f[VI]<wset_filename>\f[I]\f[R]
Decodes a wavset using smplay.
.TP
\f[B]delta\f[R] [ \f[I]\f[VI]<options>\f[I]\f[R] ] \f[I]\f[VI]<wset_filename1>\f[I]\f[R]...\f[I]\f[VI]<wset_filenameN>\f[I]\f[R]
Computes the difference between many wavsets.
The differencce between the first and second wavset is the reference
which is 100%.
The other wavsets are measured against that reference.
.TP
\f[B]link\f[R] \f[I]\f[VI]<smset>\f[I]\f[R]
This command includes all .sm files that are referred to by the smset in
one big file.
This is useful as final step when building instruments.
.SH FIELDS
.PP
For the list command, valid fields are:
.TP
\f[B]midi-note\f[R]
Midi note of the audio entry (0-127)
.TP
\f[B]channel\f[R]
Channel for the audio entry (0 is the first channel)
.TP
\f[B]filename\f[R]
Filename of the audio entry
.TP
\f[B]min-velocity\f[R]
Minimum velocity of the audio entry (0-127)
.TP
\f[B]max-velocity\f[R]
Maximum velocity of the audio entry (0-127)
.SH SEE ALSO
.PP
smenc.1, smplay.1, smstrip.1


NEWS
Overview of Changes in spectmorph-0.5.2:

* Support bpm/beat synchronization for LFO
  - new presets using beat sync LFO: Mars / Saturn
* Add WavSource custom position playback mode
* New Instruments: Sven Ah / Ih / Oh (another male human voice)
* Store data in XDG directories on Linux:
  - move ~/.spectmorph directory to $XDG_DATA_HOME/spectmorph
  - move ~/SpectMorph directory to $XDG_DOCUMENTS_DIR/SpectMorph
  - create $XDG_DOCUMENTS_DIR/SpectMorph directory when needed (on write)
  - backward compatibility: use ~/SpectMorph if it already exists
* Bump number of control inputs from 2 to 4
* Implemented midi CC control for smjack (General Purpose Controller 1..4)
* Fix crashes caused by dangling MorphOperator pointers
* Fix loading floating point wav files
* Minor fixes and cleanups

Overview of Changes in spectmorph-0.5.1:

* Add new LFO modes (saw, square, random)
* Support generic 64-bit linux binaries
  - new linux file selector (no longer needs Qt)
  - ship font for static build
* Fix crashes caused by fftw planner being used from multiple threads
* Ported all python2 code to python3
* Support midi all notes off
* Implement LV2 StateChanged
* French translation for smjack desktop file (Olivier Humbert)
* Thread race fix (JP Cimalando)
* Minor fixes and cleanups

Overview of Changes in spectmorph-0.5.0:

* Support user defined instruments
  - graphical instrument editor
  - new WavSource operator
* Make standard instrument set smaller (less download/disk usage)
* Graphical ADSR editor
* Added "SpectMorph User Manual" (online: html/pdf)
* Use different colors for active/used/unused operators
* LV2 now requires instance access feature
* Add file dialog wrapper shell scripts to work with some ardour bundles
* Integrate XML (pugixml) and ZIP (minizip) 3rd party code
* Minor fixes and cleanups

Overview of Changes in spectmorph-0.4.1:

* macOS is now supported: provide VST plugin for macOS >= 10.9
* Include instruments in source tarball and packages
* Install instruments to system-wide location
* New Instruments: Claudia Ah / Ih / Oh (female version of human voice)
* Improved tools for instrument building
  - support displaying tuning in sminspector
  - implement "smooth-tune" command for reducing vibrato from recordings
  - minor encoder fixes/cleanups
  - smlive now supports enable/disable noise
* VST plugin: fix automation in Cubase (define "effCanBeAutomated")
* UI: use Source A / Source B instead of Left Source / Right Source
* UI: update db label properly on grid instrument selection change
* Avoid exporting symbols that don't belong to the SpectMorph namespace
* Fix some LV2 ttl problems
* Fix locale related problems when using atof()
* Minor fixes and cleanups

Overview of Changes in spectmorph-0.4.0:

* Windows is now supported: provide 64-bit Windows VST plugin
* Plugin UI redesign
  - use pugl library for portability (uses GL + cairo) instead of Qt5
  - use categories for instruments
  - directly support instrument names in linear morphing
  - get rid of Qt5 dependency for libspectmorph, smjack, VST and LV2 plugins
  - UI now has "Zoom" feature to support higher DPI displays
* Use non-linear configurable new velocity -> volume mapping for midi
* New instrument: French Horn
* Improved tools for building custom instruments
  - tools are now installed by default
  - sminstbuilder files support new syntax for relative paths
  - encoder cache moved to ~/.cache/smenccache, which is created if necessary
  - use number of processors as default for jobs
* LPC/LSF support removed
* Some portability fixes for macOS (which however isn't supported yet)

Overview of Changes in spectmorph-0.3.4:

* Added optional ADSR Envelope
* Make LV2 and VST plugin stereo to allow supporting stereo in the future
* LV2 plugin description fixes
* Added about dialog to plugin/smjack UI
* Remove BEAST plugin (plugin code will be moved to BEAST)
* Fixed compilation for newer g++ >= 6 (std::fabs)
* Get rid of some malloc() calls in linear morphing

Overview of Changes in spectmorph-0.3.3:

* Added portamento:
  - VST: support MPE to perform per-voice pitch bend (can be used in Bitwig)
  - new portamento mono mode (all hosts)
* Added vibrato.
* Internal improvements:
  - better property abstraction for (non-linear) UI properties
  - updated polyphase interpolator (used for vibrato|portamento)
  - fixed a few problems when developing against spectmorph(ui) libs
  - don't link against Qt UI library when only QtCore is necessary
* Compile fixes for g++-6.3

Overview of Changes in spectmorph-0.3.2:

* Added new unison effect.
* New instruments: pan-flute, synth-saw.
* UI improvements:
  - support operator folding (to preserve screen space)
  - provide scrollbar if morph plan window height is large
  - repair operator move
* VST plugin crash fixed.
* No longer depend on BEAST/Rapicorn
  - use libsndfile for sound file I/O, added WavData API
  - refactoring, move libnobse code into SpectMorph
* Add icon/.desktop file for smjack
* Added debian package support.
* LPC/LSF morphing code updates - but now disabled by default

Overview of Changes in spectmorph-0.3.1:

* Added plugins for LV2 and VST api.
* New instruments: bassoon, cello, bass-trombone, reed-organ.
* Added different templates to get standard morph plans quickly.
* LV2|VST|JACK will start with default plan now (instead of empty plan).
* Standard instrument set location (~/.spectmorph/instruments/standard):
  - plan templates can refer to instruments in that directory without
    storing any absolute path (index will be instruments:standard)
  - in almost any case, loading instruments isn't necessary anymore
* Resize MorphPlanWindow automatically if operators are removed.
* Changed time alignment during morphing:
  - morphed sounds should starty at the beginning of the note (no extra latency)
  - Start marker for instrument notes no longer necessary
  - SpectMorphDelay plugin no longer necessary
* Some improvements for building new instruments:

lib/smwavset.cc
// Licensed GNU LGPL v2.1 or later: http://www.gnu.org/licenses/lgpl-2.1.html

#include "smwavset.hh"
#include "smoutfile.hh"
#include "sminfile.hh"
#include "smmemout.hh"

#include <map>
#include <set>

#include <assert.h>

using std::vector;
using std::string;
using std::map;
using std::set;

using namespace SpectMorph;

Error
WavSet::save (const string& filename, bool embed_models)
{
  OutFile of (filename.c_str(), "SpectMorph::WavSet", SPECTMORPH_BINARY_FILE_VERSION);
  if (!of.open_ok())
    {
      fprintf (stderr, "error: can't open output file '%s'.\n", filename.c_str());
      exit (1);
    }

  of.write_string ("name", name);
  of.write_string ("short_name", short_name);

  for (size_t i = 0; i < waves.size(); i++)
    {
      of.begin_section ("wave");
      of.write_int ("midi_note", waves[i].midi_note);
      of.write_int ("channel", waves[i].channel);
      of.write_int ("velocity_range_min", waves[i].velocity_range_min);
      of.write_int ("velocity_range_max", waves[i].velocity_range_max);
      of.write_string ("path", waves[i].path);
      if (waves[i].audio)
        {
          vector<unsigned char> data;

          waves[i].audio->save (MemOut::open (&data));
          of.write_blob ("audio", data.data(), data.size());
        }
      else if (embed_models)
        {
          FILE *in = fopen (waves[i].path.c_str(), "rb");
          if (in)
            {
              vector<unsigned char> data;
              vector<unsigned char> buffer (1024);
              size_t len;

              do
                {
                  len = fread (buffer.data(), 1, buffer.size(), in);
                  if (len > 0)
                    data.insert (data.end(), buffer.begin(), buffer.begin() + len);
                }
              while (len > 0);

              if (ferror (in))
                {
                  fprintf (stderr, "wavset save: error reading file: %s\n", waves[i].path.c_str());
                }
              else
                {
                  of.write_blob ("audio", data.data(), data.size());
                }
              fclose (in);
            }
          else
            {
              fprintf (stderr, "wavset save: missing file: %s\n", waves[i].path.c_str());
            }
        }
      of.end_section();
    }
  return Error::Code::NONE;
}

Error
WavSet::load (const string& filename, AudioLoadOptions load_options)
{
  clear();        // delete old contents (if any)

  map<string, Audio *> blob_map;

  WavSetWave *wave = NULL;

  InFile  ifile (filename);
  string section;

  if (!ifile.open_ok())
    return Error::Code::FILE_NOT_FOUND;

  if (ifile.file_type() != "SpectMorph::WavSet")
    return Error::Code::FORMAT_INVALID;

  if (ifile.file_version() != SPECTMORPH_BINARY_FILE_VERSION)
    return Error::Code::FORMAT_INVALID;

  while (ifile.event() != InFile::END_OF_FILE)
    {
      if (ifile.event() == InFile::BEGIN_SECTION)
        {
          assert (section == "");
          section = ifile.event_name();

          if (section == "wave")
            {
              assert (wave == NULL);
              wave = new WavSetWave();
            }
        }
      else if (ifile.event() == InFile::END_SECTION)
        {
          if (section == "wave")
            {
              assert (wave);

              waves.push_back (*wave);
              delete wave;
              wave = NULL;
            }

          assert (section != "");
          section = "";
        }
      else if (ifile.event() == InFile::INT)
        {
          if (section == "wave")
            {
              if (ifile.event_name() == "midi_note")
                {
                  assert (wave);
                  wave->midi_note = ifile.event_int();
                }
              else if (ifile.event_name() == "channel")
                {
                  assert (wave);
                  wave->channel = ifile.event_int();
                }
              else if (ifile.event_name() == "velocity_range_min")
                {
                  assert (wave);
                  wave->velocity_range_min = ifile.event_int();
                }
              else if (ifile.event_name() == "velocity_range_max")
                {
                  assert (wave);
                  wave->velocity_range_max = ifile.event_int();
                }
              else
                printf ("unhandled int %s %s\n", section.c_str(), ifile.event_name().c_str());
            }
          else
            assert (false);
        }
      else if (ifile.event() == InFile::STRING)
        {
          if (section == "wave")
            {
              if (ifile.event_name() == "path")
                {
                  assert (wave);
                  wave->path = ifile.event_data();
                }
              else
                printf ("unhandled string %s %s\n", section.c_str(), ifile.event_name().c_str());
            }
          else if (section == "")
            {
              if (ifile.event_name() == "name")
                {
                  name = ifile.event_data();
                }
              else if (ifile.event_name() == "short_name")
                {
                  short_name = ifile.event_data();
                }
              else
                printf ("unhandled global string %s\n", ifile.event_name().c_str());
            }
          else
            assert (false);
        }
      else if (ifile.event() == InFile::BLOB)
        {
          if (section == "wave")
            {
              if (ifile.event_name() == "audio")
                {
                  assert (wave);
                  assert (!wave->audio);

                  GenericInP blob_in = ifile.open_blob();

                  wave->audio = new Audio();
                  wave->audio->load (blob_in, load_options);

                  blob_map[ifile.event_blob_sum()] = wave->audio;
                }
              else
                printf ("unhandled string %s %s\n", section.c_str(), ifile.event_name().c_str());
            }
          else
            assert (false);
        }
      else if (ifile.event() == InFile::BLOB_REF)
        {
          if (section == "wave")
            {
              if (ifile.event_name() == "audio")
                {
                  assert (wave);
                  assert (!wave->audio);

                  wave->audio = blob_map[ifile.event_blob_sum()];

                  assert (wave->audio);
                }
              else
                printf ("unhandled string %s %s\n", section.c_str(), ifile.event_name().c_str());
            }
          else
            assert (false);
        }
      else if (ifile.event() == InFile::READ_ERROR)
        {
          return Error::Code::PARSE_ERROR;
        }
      else
        {
          assert (false);
        }
      ifile.next_event();
    }
  return Error::Code::NONE;
}

WavSetWave::WavSetWave()
{
  audio = NULL;
  channel = 0;
  midi_note = -1;
  velocity_range_min = 0;
  velocity_range_max = 127;
}

WavSetWave::~WavSetWave()
{
  // delete audio; // <- don't do it here, because of the constructor/destructor calls in the vector
}

void
WavSet::clear()
{
  set<Audio *> to_delete;

  for (vector<WavSetWave>::iterator wi = waves.begin(); wi != waves.end(); wi++)
    {
      if (wi->audio)
        {
          to_delete.insert (wi->audio);
          wi->audio = NULL;
        }
    }

  /* This is a little piece of extra logic that prevents deleting one Audio entry twice;
     the same entry can be reused in one WavSet multiple times (for instance for different
     velocity levels) - and while BLOB_REFs care for reusing the same object in this case,
     we need to ensure here we don't delete the same object twice */
  for (set<Audio *>::const_iterator di = to_delete.begin(); di != to_delete.end(); di++)
    delete *di;

  // now that everything has been delete-d, we can reset the waves vector
  waves.clear();
}

WavSet::~WavSet()
{
  clear();
}
